### PR TITLE
Fix rebuild voxel shape cost too much time, close #314

### LINKED ISSUE #314: [1.20.1] Game frozen for few seconds when linking/unlinking a drawer to a controller
**Functional Storage Version:** 1.20.1-1.2.10

**Titanium Version:** 3.8.32

I want to link a new drawer to my controller which has been connected with **hundreds of drawers**. But game is froze for a few seconds each time I try do it. And it only happen on **client-side**, the **server-side** is normal. (I mean, not cause server tps down rapidly.)

There is spark profiler report snapshot.

I see the source code, `rebuildShapes` cost too much time, can you optimize it? or is there other solution?
[rebuildShapes method in source code](https://github.com/Buuz135/FunctionalStorage/blob/e3f1b16ecb6e9487188f2308fd3aa563b3a6273d/src/main/java/com/buuz135/functionalstorage/util/ConnectedDrawers.java#L84)

Also, I find the same situation when a LinkingTool in my hand for a while.
[the reason in source code](https://github.com/Buuz135/FunctionalStorage/blob/e3f1b16ecb6e9487188f2308fd3aa563b3a6273d/src/main/java/com/buuz135/functionalstorage/client/ControllerRenderer.java#L90)

### PR DESCRIPTION
**Pull request by SeaLoong**
I tested the code in my case(#314). I hardly feel any frozen.

In addition, it can decrease the time cost down to less than 100ms(in my case), just replace `Shapes.join` with `Shapes.joinUnoptimized`

### CODE REVIEW

File: src/main/java/com/buuz135/functionalstorage/util/ConnectedDrawers.java [hunk 1]
```diff
@@ -84,8 +84,23 @@ public void rebuild() {
...
+                    var shape2 = currentShapes.get(i + 1);
+                    // Notice: DO NOT use `Shapes.join`, `join` function will 'optimize' the joined shape, it traversals all exist boxes each time.
+                    mergedShape = Shapes.joinUnoptimized(shape1, shape2, BooleanOp.OR);
+                }
+                mergedShapes.add(mergedShape);
+            }
+            currentShapes = mergedShapes;
+        } while (currentShapes.size() > 1);
+        if (currentShapes.size() == 1) {
+            this.cachedVoxelShape = Shapes.joinUnoptimized(this.cachedVoxelShape, currentShapes.get(0), BooleanOp.OR);
```
Comment (Matyrobbrt) [inline src/main/java/com/buuz135/functionalstorage/util/ConnectedDrawers.java:103]: What's the point of this hard to follow loop? Just replace the current `join` with `joinUnoptimized` and optimise the shape at the end.

**Review by Matyrobbrt — CHANGES_REQUESTED** [review]

File: src/main/java/com/buuz135/functionalstorage/util/ConnectedDrawers.java (hunk 1 above)
Comment (SeaLoong) [reply src/main/java/com/buuz135/functionalstorage/util/ConnectedDrawers.java:103]: Well, I just try joining voxel shapes by chunk,  like **Merge Sort**. I don't read the implement of Shape class. But it works, it can significantly reduce the time cost.

**Comment by SeaLoong** [comment]
## Performance Difference
- The controller has connected with 16 x 20 x 16 Drawers and 16 x 20 x 16 Extensions

### Replace `Shapes.join` with `Shapes.joinUnoptimized`

https://github.com/user-attachments/assets/0d47b3e4-7b50-4120-b094-75afed4c3948

There is about 1s game frozen in this test case.

### Using the loop implement, also using replaced join function

https://github.com/user-attachments/assets/e3e526de-f9a7-44f6-b4ec-a0c1348ceb83

You can see there is still a bit game frozen, in this test case. But it is hardly perceivable.
